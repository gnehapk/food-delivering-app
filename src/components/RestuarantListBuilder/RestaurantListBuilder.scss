@import '../../color-variables.scss';

.restaurant-list {
  display: flex;
  flex-wrap: wrap;
  margin: 0 4rem;
  justify-content: center;
  flex-direction: column;

  @media only screen and (min-width: 768px) {
    flex-direction: row;
  }
}

.restaurant-count {
  display: none;

  @media only screen and (min-width: 768px) {
    display: block;
    margin: 2rem 0 0 20rem;
    color: $color-dark-grey;
  }
}