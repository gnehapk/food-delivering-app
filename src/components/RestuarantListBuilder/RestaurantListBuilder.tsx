import React from 'react';
import Spinner from 'react-bootstrap/Spinner';

import Header from '../Header';
import getRestaurantList from '../../assets/restaurant-list';
import Restaurant from '../Restaurant';

import FilterSection from '../FilterSection';

import './RestaurantListBuilder.scss';

interface ObjectLiteral {
  [key: string]: number;
}

const RestaurantListBuilder: React.FC = () => {
  const [loadingState, setLoadingState] = React.useState<boolean>(false);
  const [restaurantList, setRestaurantList] = React.useState<RestaurantType[]>(
    []
  );
  const [filters, setFilters] = React.useState<ObjectLiteral>({});
  const [location, setLocation] = React.useState<string>('');

  React.useEffect(() => {
    setLoadingState(true);
    //actual request will look like this
    // http.get('url')
    //   .then(list =>  {
    //     console.log(list);
    //   }).catch(err => console.log(err.message));

    // hardcoding for now
    const { location, restaurants } = getRestaurantList();
    setLocation(location);
    setRestaurantList(restaurants);
    setLoadingState(false);

    let tempMap: ObjectLiteral = {};
    for(const restaurant of restaurants) {
      
      for(const tag of restaurant.tags) {
        if(tempMap.hasOwnProperty(tag)) {
          tempMap[tag] += 1;
        } else {
          tempMap[tag] = 1;
        }
      }
    }
    setFilters(tempMap);
  }, []);

  const doFilter = (event: React.MouseEvent<HTMLLIElement>, tag: string) => {
    let sortedList = [];

    if(!tag) {
      sortedList = getRestaurantList().restaurants;
    } else {
      sortedList = restaurantList.filter((r: RestaurantType) => {
        return r.tags.includes(tag);
      });
    }
    setRestaurantList(sortedList);
  };

  
  return (
    <div>
      <Header location={location} />
      {loadingState && <Spinner animation="border" variant="info" />}
      <FilterSection tags={filters} doFilter={doFilter}/>
      {!loadingState && (
        <div>
          <div className="restaurant-count">
            {restaurantList.length} Restaurants
          </div>
          <div className="restaurant-list">
            {restaurantList.map((restaurant: RestaurantType) => (
              <Restaurant key={restaurant.id} details={restaurant} />
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export type RestaurantType = {
  id: string;
  name: string;
  image: string;
  url: string;
  price: number;
  tags: string[];
};

export default RestaurantListBuilder;
