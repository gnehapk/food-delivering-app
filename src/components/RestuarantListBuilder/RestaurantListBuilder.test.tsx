import React from 'react';
import TestRenderer from 'react-test-renderer';
import RestaurantListBuilder from './RestaurantListBuilder';

describe('RestaurantListBuilder', () => {
  it('should render correctly', () => {
    const tree = TestRenderer.create(<RestaurantListBuilder />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

