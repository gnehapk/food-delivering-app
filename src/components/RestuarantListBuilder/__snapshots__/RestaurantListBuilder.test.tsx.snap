// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantListBuilder should render correctly 1`] = `
<div>
  <div
    className="header"
  >
    <div>
      <p
        className="location-text"
      >
        Location
      </p>
      <p
        className="location-name"
      >
        
      </p>
    </div>
    <div>
      <button
        className="location-btn btn btn-light"
        disabled={false}
        type="button"
      >
        Change Location
      </button>
    </div>
  </div>
  <div>
    <div
      className="restaurant-count"
    >
      0
       Restaurants
    </div>
    <div
      className="restaurant-list"
    />
  </div>
</div>
`;
